libopeniscsiusr/idbm.c

```diff
@@ -0,0 +1,699 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Gris Ge <fge@redhat.com>
+ */
+
+/* The code below is modified from usr/idbm.c which licensed like below:
+ *
+ * iSCSI Discovery Database Library
+ *
+ * Copyright (C) 2004 Dmitry Yusupov, Alex Aizman
+ * Copyright (C) 2006 Mike Christie
+ * Copyright (C) 2006 Red Hat, Inc. All rights reserved.
+ * maintained by open-iscsi@@googlegroups.com
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published
+ * by the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * See the file COPYING included with this distribution for more details.
+ *
+ */
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE			/* For strerror_r() */
+#endif
+
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <string.h>
+#include <inttypes.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <ctype.h>
+
+#include "libopeniscsiusr/libopeniscsiusr_common.h"
+
+#include "context.h"
+#include "idbm.h"
+#include "misc.h"
+#include "idbm_fields.h"
+#include "iface.h"
+#include "version.h"
+
+#define TYPE_STR	2
+#define TYPE_UINT8	3
+#define TYPE_UINT16	4
+#define TYPE_UINT32	5
+#define MAX_KEYS	256   /* number of keys total(including CNX_MAX) */
+#define NAME_MAXVAL	128   /* the maximum length of key name */
+#define VALUE_MAXVAL	256   /* the maximum length of 223 bytes in the RFC. */
+/* ^ MAX_KEYS, NAME_MAXVAL and VALUE_MAXVAL are copied from usr/idbm.h
+ * The RFC 3720 only said:
+ *	If not otherwise specified, the maximum length of a simple-value (not
+ *	its encoded representation) is 255 bytes, not including the delimiter
+ *	(comma or zero byte).
+ */
+
+#define OPTS_MAXVAL	8
+
+#define IDBM_HIDE	0    /* Hide parameter when print. */
+#define IDBM_SHOW	1    /* Show parameter when print. */
+#define IDBM_MASKED	2    /* Show "stars" instead of real value when print */
+
+#define ISCSI_BEGIN_REC	"# BEGIN RECORD "ISCSI_VERSION_STR
+#define ISCSI_END_REC	"# END RECORD"
+
+#ifndef LOCK_DIR
+#define LOCK_DIR		"/var/lock/iscsi"
+#endif
+#define LOCK_FILE		LOCK_DIR"/lock"
+#define LOCK_WRITE_FILE		LOCK_DIR"/lock.write"
+
+#define _rec_str(_key, _recs, _org, _name, _show, _n, _mod) \
+do { \
+	_recs[_n].type = TYPE_STR; \
+	strncpy(_recs[_n].name, _key, NAME_MAXVAL); \
```

Comment:
If you give the entire buffer to use for `strncpy` and the source string is >= length the destination string will not be null terminated.

Reply:
Thanks for pointing that out. Fixed by a macro `_strncpy()` setting the null terminator.